stan/math/rev/functor/operands_and_partials.hpp

```diff
@@ -194,6 +234,35 @@ class ops_partials_edge<double, Op, require_eigen_st<is_var, Op>> {
         = this->partials_;
   }
   int size() { return this->operands_.size(); }
+  std::tuple<> container_operands() { return std::tuple<>(); }
+  std::tuple<> container_partials() { return std::tuple<>(); }
+};
+
+template <typename Op>
+class ops_partials_edge<double, var_value<Op>, require_eigen_t<Op>> {
+ public:
+  using partials_t = plain_type_t<Op>;
+  partials_t partials_;                       // For univariate use-cases
+  broadcast_array<partials_t> partials_vec_;  // For multivariate
+  explicit ops_partials_edge(const var_value<Op>& ops)
+      : partials_(partials_t::Zero(ops.vi_->rows(), ops.vi_->cols())),
+        partials_vec_(partials_),
+        operands_(ops) {}
+
+ private:
+  template <typename, typename, typename, typename, typename, typename>
+  friend class stan::math::operands_and_partials;
+  const var_value<Op>& operands_;
+
+  void dump_operands(vari** varis) {}
+  void dump_partials(double* partials) {}
+  int size() { return 0; }
+  std::tuple<const var_value<Op>&> container_operands() {
+    return std::forward_as_tuple(operands_);
+  }
+  std::tuple<partials_t&&> container_partials() {
+    return std::forward_as_tuple(std::move(partials_));
```

Comment:
Could you walk me through the forwarding/moving here?

Reply:
I am modifying this a bit. Now partials in edges are already allocated on AD stack, so there is no need to do a deep copy when putting them on AD stack in `precomputed_gradients`. This call to `std::move` here is not necessary anymore so I removed it. 

For both operands and partials `forward_as_tuple` constructs a tuple containing a lvalue reference. Than nothing interesting happens until the constructor of `precomputed_gradients`. In the constructor a call to `to_AD_stack` copies operands to stack. Partials are already on stack so `to_AD_stack` just makes a shallow copy (equivalent to move, except the original object is unchanged).